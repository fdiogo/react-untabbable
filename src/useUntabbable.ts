import tabbable from "tabbable";
import { useEffect } from "react";

type Descriptor = {
	element: HTMLElement;
	hadTabindex: boolean;
	tabindex: string | null;
};

type Options = {
    /** 
     * Disable the component and restore the tabbable behaviour.
     */
    disabled?: boolean,
    /**
     * If it should also disable the container received as a ref.
     */
	includeContainer?: boolean,
};

function createDescriptor(element: HTMLElement) : Descriptor {
    return {
        element,
        hadTabindex: element.hasAttribute('tabindex'),
        tabindex: element.getAttribute('tabindex'),
    };
}

function setElementUntabbable(element: HTMLElement) {
    element.setAttribute('tabindex', '-1');
}

function restoreTabbable(descriptor: Descriptor) {
    const { element, hadTabindex, tabindex } = descriptor;

    if (!hadTabindex) {
        element.removeAttribute('tabindex');
    } else {
        element.setAttribute('tabindex', `${tabindex}`);
    }
}

function useUntabbable(ref: React.RefObject<HTMLElement>, options: Options = {} ) {
    const { disabled = false, includeContainer = false } = options;

    useEffect(() => {
        if (disabled || !ref.current) {
            return;
        }

        const tabbableElements = tabbable(ref.current, { includeContainer });
        const descriptors = tabbableElements.map(createDescriptor);

        tabbableElements.forEach(setElementUntabbable);

        return () => {
            descriptors.forEach(restoreTabbable);
        };
    }, [ref, disabled, includeContainer]);
}

export default useUntabbable;
