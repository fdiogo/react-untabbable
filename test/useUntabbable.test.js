import React from 'react';
import { render, cleanup } from '@testing-library/react';

import useUntabbable from '../src/useUntabbable';

afterEach(cleanup);

// We mock tabbable here since what we are really interested on
// testing is the toggling of the tabindex.
jest.mock('tabbable', () => (element, options) => {
    const tabbableElements = [];

    if(options.includeContainer) {
        tabbableElements.push(element);
    }

    tabbableElements.push(...element.children);

    return tabbableElements;
});

const HookWrapper = (props) => {
    const { children, targetRef, options } = props;

    useUntabbable(targetRef, options)

    return children;
};

describe('useUntabbable', () => {
    it('should set the focusable children with tabindex=-1', () => {
        const ref = React.createRef();

        const { container, getByText } = render(
            <HookWrapper targetRef={ ref }>
                <div ref={ ref }>
                    <button>First</button>
                    <button>Second</button>
                </div>
            </HookWrapper>
        );

        const firstButton = getByText('First');
        const secondButton = getByText('Second');

        expect(container.firstChild.getAttribute("tabindex")).not.toBe("-1");
        expect(firstButton.getAttribute("tabindex")).toBe("-1");
        expect(secondButton.getAttribute("tabindex")).toBe("-1");
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should restore the previous tabindex when disabled', () => {
        const ref = React.createRef();

        const { container, getByText} = render(
            <HookWrapper targetRef={ ref }>
                <div ref={ ref }>
                    <button tabIndex="2">First</button>
                    <button>Second</button>
                </div>
            </HookWrapper>
        );

        const firstButton = getByText('First');
        const secondButton = getByText('Second');

        expect(firstButton.getAttribute("tabindex")).toBe("-1");
        expect(secondButton.getAttribute("tabindex")).toBe("-1");

        render(
            <HookWrapper targetRef={ ref } options={ { disabled: true } }>
                <div ref={ ref }>
                    <button tabIndex="2">First</button>
                    <button>Second</button>
                </div>
            </HookWrapper>,
        { container });

        expect(firstButton.getAttribute("tabindex")).toBe("2");
        expect(secondButton.getAttribute("tabindex")).toBe(null);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should also disable the container when `includeContainer` is true', () => {
        const ref = React.createRef();

        const { container, getByText } = render(
            <HookWrapper targetRef={ ref } options={ { includeContainer: true } }>
                <div ref={ ref }>
                    <button>First</button>
                    <button>Second</button>
                </div>
            </HookWrapper>
        );

        const firstButton = getByText('First');
        const secondButton = getByText('Second');

        expect(container.firstChild.getAttribute("tabindex")).toBe("-1");
        expect(firstButton.getAttribute("tabindex")).toBe("-1");
        expect(secondButton.getAttribute("tabindex")).toBe("-1");
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should not alter the tabindex when `disabled` is true', () => {
        const ref = React.createRef();

        const { container, getByText } = render(
            <HookWrapper targetRef={ ref } options={ { disabled: true } }>
                <div ref={ ref }>
                    <button>First</button>
                    <button>Second</button>
                </div>
            </HookWrapper>
        );

        const firstButton = getByText('First');
        const secondButton = getByText('Second');

        expect(container.firstChild.getAttribute("tabindex")).not.toBe("-1");
        expect(firstButton.getAttribute("tabindex")).not.toBe("-1");
        expect(secondButton.getAttribute("tabindex")).not.toBe("-1");
        expect(container.firstChild).toMatchSnapshot();
    });
});