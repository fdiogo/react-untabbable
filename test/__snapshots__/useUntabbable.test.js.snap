// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useUntabbable should also disable the container when \`includeContainer\` is true 1`] = `
<div
  tabindex="-1"
>
  <button
    tabindex="-1"
  >
    First
  </button>
  <button
    tabindex="-1"
  >
    Second
  </button>
</div>
`;

exports[`useUntabbable should not alter the tabindex when \`disabled\` is true 1`] = `
<div>
  <button>
    First
  </button>
  <button>
    Second
  </button>
</div>
`;

exports[`useUntabbable should restore the previous tabindex when disabled 1`] = `
<div>
  <button
    tabindex="2"
  >
    First
  </button>
  <button>
    Second
  </button>
</div>
`;

exports[`useUntabbable should set the focusable children with tabindex=-1 1`] = `
<div>
  <button
    tabindex="-1"
  >
    First
  </button>
  <button
    tabindex="-1"
  >
    Second
  </button>
</div>
`;
